reactor-netty-core/src/main/java/reactor/netty/ReactorNetty.java

```diff
@@ -721,42 +722,41 @@ public void onStateChange(Connection connection, State newState) {
 	 * An appending write that delegates to its origin context and append the passed
 	 * publisher after the origin success if any.
 	 */
-	static final class OutboundThen implements NettyOutbound {
+	static final class OutboundThen extends AtomicBoolean implements NettyOutbound {
 
 		final NettyOutbound source;
 		final Mono<Void> thenMono;
 
 		static final Runnable EMPTY_CLEANUP = () -> {};
 
-
 		OutboundThen(NettyOutbound source, Publisher<Void> thenPublisher) {
 			this(source, thenPublisher, EMPTY_CLEANUP);
 		}
 
+		// This construction is used only with ChannelOperations#sendObject
+		// The implementation relies on Netty's promise that Channel#writeAndFlush will release the buffer on success/error
```

Comment:
It would be worth to add another line comment that the release hear happens only in case when we are sure that the processing doesn't delegate to Netty because of some failure before the exchange can be continued in the `thenPublisher` - which in a regular case takes ownership of the buffer.